test/suite/test_backup09.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env python
+#
+# Public Domain 2014-2018 MongoDB, Inc.
+# Public Domain 2008-2014 WiredTiger, Inc.
+#
+# This is free and unencumbered software released into the public domain.
+#
+# Anyone is free to copy, modify, publish, use, compile, sell, or
+# distribute this software, either in source code form or as a compiled
+# binary, for any purpose, commercial or non-commercial, and by any
+# means.
+#
+# In jurisdictions that recognize copyright laws, the author or authors
+# of this software dedicate any and all copyright interest in the
+# software to the public domain. We make this dedication for the benefit
+# of the public at large and to the detriment of our heirs and
+# successors. We intend this dedication to be an overt act of
+# relinquishment in perpetuity of all present and future rights to this
+# software under copyright law.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# test_backup09.py
+#   Verify opening a backup cursor forces a log file switch.
+#
+
+import os, shutil
+import wiredtiger, wttest
+from wtscenario import make_scenarios
+
+class test_backup09(wttest.WiredTigerTestCase):
+    conn_config = 'config_base=false,create,log=(enabled)'
+    uri = 'table:coll1'
+    dir = 'backup.dir'
+
+    types = [
+        ('checkpoint', {'checkpoint': True}),
+        ('no_checkpoint', {'checkpoint': False}),
+    ]
+    scenarios = make_scenarios(types)
+
+    def data_and_start_backup(self):
+        self.session.create(self.uri, 'key_format=i,value_format=i')
+
+        cursor = self.session.open_cursor(self.uri)
+        doc_id = 0
+
+        for i in range(10):
+            doc_id += 1
+            cursor[doc_id] = doc_id
+
+        if self.checkpoint:
+            self.session.checkpoint()
+
+        for i in range(10):
+            doc_id += 1
+            cursor[doc_id] = doc_id
+
+        last_doc_in_backup = doc_id
+        self.assertEqual(1, len(filter(lambda x: x.startswith('WiredTigerLog.'),
+                                        os.listdir('.'))))
+        backup_cursor = self.session.open_cursor('backup:')
+        # Opening a backup cursor will immediately force a new log file.
```

Comment:
I thought the flag is a lazy file creation when the next log record gets written and we had discussed possibly adding a log printf to the test or just checking after adding more data. What log record is getting written?

Reply:
We talked about this and it resulted in add a forced a log write after setting the flag: https://github.com/wiredtiger/wiredtiger/commit/8b423cbcf1b317bf617ff54f787e1d7f3b544bd6#diff-f948a1575eef3a5b686ffed3802c2b5aR185

Before that change, I was seeing:
1. Perform write A
2. Assert 1 log file
3. Open a backup cursor
4. Perform write B
5. Assert 2 log files

Which was failing the test. I had expected writes concurrent with opening a backup cursor may or may not appear in the log files to copy, but not writes causally after the backup cursor.

I now realized though that this feature/test in particular wants to assert that taking a checkpoint between steps 3 and 4 results in a new log file. When I remove the forced log write and add the checkpoint call, the two log files become visible via the OS right after the checkpoint without any additional "flushing" calls.

The only part I don't completely understand is that taking a checkpoint will respect the `WT_LOG_FORCE_NEWFILE` flag, but a "typical write" does not. Does it have something to do with log slots and which operations can dispatch an active, not full slot?

My next patch will remove the forced log write and add a checkpoint call in the test.

Reply:
FWIW, removing the line that forces the log write fixed some scenarios in the txn test (but not all).

Reply:
Ah! Right! Forcing out earlier writes and forcing a log slot switch causes the flag to be noticed when setting up the next, new slot. You cannot remove the call to {{wt_log_force_write}}. Its purpose is to force out earlier records in the slot's log buffer memory and write(2) them out to the log file. It is NOT to force some new record to be written into the log.

You must force out those records before returning the backup cursor so that all existing records in the log are written to the OS before the backup cursor returns that filename to be copied.

Reply:
Are you saying that, today, opening up a backup cursor does not force out earlier records?

Reply:
AFAICT, that is correct. That is a bug, just realized now, and fixed with this.